src/lib.rs

```diff
@@ -0,0 +1,250 @@
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new();
+    /// let c2 = c.push(1);
+    /// let c3 = c2.push(2);
+    /// assert_eq!(c3.iter().cloned().collect::<Vec<_>>(), [2, 1]);
+    /// ```
+    pub fn push(&self, val: T) -> Cactus<T> {
+        Cactus {
+            node: Some(Rc::new(Node{val, parent: self.node.clone()}))
+        }
+    }
+
+    /// Return this cactus stack node's parent node or `None` if this cactus stack is empty.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new();
+    /// let c2 = c.push(1);
+    /// assert_eq!(c.pop(), None);
+    /// assert_eq!(c2.val(), Some(&1));
+    /// assert_eq!(c2.pop().unwrap(), Cactus::new());
+    /// ```
+    pub fn pop(&self) -> Option<Cactus<T>> {
+        self.node.as_ref().map(|n| Cactus{node: n.parent.clone()} )
+    }
+
+    /// Return a reference to this cactus stack node's value or `None` if this cactus stack is
+    /// empty.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new().push(1);
+    /// assert_eq!(c.val(), Some(&1));
+    /// assert_eq!(c.pop().unwrap().val(), None);
+    /// ```
+    pub fn val(&self) -> Option<&T> {
+        self.node.as_ref().map(|n| &n.val)
+    }
+
+    /// Return an iterator over this cactus stack's values. Note that the iterator produces values
+    /// starting from this node and then walking up towards the root.
+    ///
+    /// # Examples
+    /// ```
+    /// use cactus::Cactus;
+    /// let c = Cactus::new().push(1).push(2).push(3);
+    /// assert_eq!(c.iter().cloned().collect::<Vec<_>>(), [3, 2, 1]);
+    /// ```
+    pub fn iter(&self) -> CactusIter<T> {
+        CactusIter{next: self.node.as_ref()}
+    }
+}
+
+/// An iterator over a `Cactus` stack. Note that the iterator produces values starting from this
+/// node and then walking up towards the root.
+pub struct CactusIter<'a, T> where T: 'a {
+    next: Option<&'a Rc<Node<T>>>
+}
+
+impl<'a, T> Iterator for CactusIter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.take().map(|n| {
+            self.next = n.parent.as_ref();
+            &n.val
+        })
+    }
+}
+
+impl<T: PartialEq> PartialEq for Cactus<T> {
+    fn eq(&self, other: &Cactus<T>) -> bool {
+        if self.len() != other.len() {
+            return false;
+        }
+        self.iter().zip(other.iter()).all(|(x, y)| x == y)
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for Cactus<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        try!(write!(f, "Cactus["));
+        for (i, x) in self.iter().enumerate() {
+            if i > 0 {
+                try!(write!(f, ", "));
+            }
+            try!(write!(f, "{:?}", x));
+        }
+        write!(f, "]")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_simple() {
```

Comment:
Would it be better to split this larger test into smaller ones for: e.g. `pop`, `is_empty`, like you did for `iter`?

I guess you could keep the big test too.